src/coreclr/vm/jithelpers.cpp

```diff
@@ -2858,45 +2859,15 @@ HCIMPL2(Object*, JIT_Box, CORINFO_CLASS_HANDLE type, void* unboxedData)
         FCThrow(kNullReferenceException);
     }
 
-    // Use the relevant allocation context to try to do a trail allocation
-    // without calling into the GC.
-    gc_alloc_context* pAllocContext;
-    if (GCHeapUtilities::UseThreadAllocationContexts())
-    {
-        pAllocContext = GetThread()->GetAllocContext();
-    }
-    else
-    {
-        pAllocContext = &g_global_alloc_context;
-    }
-
-    uint32_t size = clsHnd.GetSize();
-
-    if (!GCHeapUtilities::UseThreadAllocationContexts())
-    {
-        GlobalAllocLock::AcquireLock(&g_global_alloc_lock);
-    }
+    OBJECTREF newobj = TryBoxFromTrialAllocation(pMT, unboxedData);
```

Comment:
Unfortunately, this won't work well. There are number of other conditions that must be checked to select the right allocation helper: profilers, GC stress, special alignment requirements, ... : https://github.com/dotnet/runtime/blob/5c8bb402e6a8274e8135dd00eda2248b4f57102f/src/coreclr/vm/jitinterface.cpp#L5878-L5963

Computing the right way to allocate an object is complicated. It means that it has to be cached somewhere. For JITed code, it is "cached" by burning pointers to the allocation helpers into the code.

I think it would work best to cache the box information similar how we are caching the Activator.CreateInstance information or RuntimeHelpers.GetUninitializedObject information. It can look something like this: https://github.com/jkotas/runtime/commit/83a4311f86f23e1158aa34ea364d779d0fd2fef7

The GenericCache on RuntimeType may need a bit of refactoring for this. There is 
- If we are caching a single thing, make the generic cache point to it directly like it does today.
- If we are caching multiple things, I would introduce an intermediate type that can point to all things that can be possibly cached.

Also, boxing should not trigger static constructors. It makes it different from RuntimeHelpers.GetUninitializedObject that does trigger precise static constructors. It would be nice to add a test for that if we do not have one already.

Reply:
It looks like for boxing, the checks are much simpler than the generic allocation path. I'll see what I can do here to get something straightforward.

Reply:
My preference would be to avoid replicating/inling the logic that selects the right fast allocation path. It should help us with HELPER_FRAME removal effort.

Reply:
I'm planning on taking a similar path as the implementation for `new()`.